### PR TITLE
Display DIscord Avatars

### PR DESCRIPTION
**Pull request by haotian2006**
Created a system that can retrieve a person's discord avatar and username from userid. 

image

image

probably gonna make changes to how it's displayed and change how the data is stored since right now its just stored in the config 

image

### CODE REVIEW

**Comment by haotian2006** [comment]
Changed it so you can only put 1 author field that is either discord github or neither

"discord: userid"
image

"github: userid"
image

"Name"
image

**Comment by cyrus01337** [comment]
Great work on this so far, I'll take a look at lunch/after work.

File: src/content/config.ts [hunk 1]
```diff
@@ -5,7 +5,12 @@ const documents = defineCollection({
     schema: zod.object({
         title: zod.string(),
         search: zod.boolean().default(true),
-        author: zod.string().default("Anonymous"),
+
+        author: zod.string().optional(),
+
+        author_data: zod.record(zod.unknown()).optional(),
```
Comment (cyrus01337) [inline src/content/config.ts:11]: ## TL;DR: `config.ts` (bad name by Astro, will look into) defines properties in MDX files. `author_data` isn't shown to be used anywhere but `src/pages/[...slug].astro`, so it should be a type in that file or a different file.

Admittedly, `config.ts` isn't the most descriptive name that Astro has picked. This file determines what you an expect in MDX files within [`src/content/documents`](https://github.com/haotian2006/MyDocs/tree/Integrate-DiscordAPI/src/content/documents), so `author_data` which is not defined in other MDX files and is exclusively used in `src/pages/[...slug].astro` should be a type and either be placed there or shoved into a separate file altogether. I'll see if we can't choose a more apt name separately.

File: src/pages/[...slug].astro [hunk 2]
```diff
@@ -46,13 +125,24 @@ const extractUsername = (string:string) => {
...
             <p class="mr-4">by:</p> 
-            <img src={`${entry.data.author}.png`} alt={"author"} width={25} height={25} class="mr-2" />
-            <a href={entry.data.author} target="_blank" rel="noopener noreferrer">
-              {extractUsername(entry.data.author)}
-            </a>
+
+            
+            <img src={ (entry.data.author_data as {pfp:string}).pfp } class="w-8 h-8 rounded-full mr-2" />
+
+            { (entry.data.author_data as {url:string}).url &&(
```
Comment (cyrus01337) [inline src/pages/[...slug].astro:135]: The initial opening parentheses `(` was never closed, causing a syntax error. Please remove.

File: src/pages/[...slug].astro [hunk 3]
```diff
@@ -1,38 +1,117 @@
...
+
 import { getCollection } from "astro:content";
 
 import Layout from "@/layouts/base.astro";
-import { sys } from "typescript";
 
-export async function getStaticPaths() {
-    const documents = await getCollection("documents");
+export const discordCache = new Map<string,{username:string,pfp:string}|undefined>();
+export const githubCache = new Map<string,{username:string,pfp:string,url:string}|undefined>();
```
Comment (cyrus01337) [inline src/pages/[...slug].astro:8]: Use `T | null` as `Map<T>.get(...)` returns `T | undefined` - `null` should be used as the special-case value. This is detailed in a separate comment further into the file.

File: src/pages/[...slug].astro [hunk 4]
```diff
@@ -1,38 +1,117 @@
...
 
-    const documentPaths = documents.map(entry => ({
-    params: {
-      slug: entry.slug === 'index' ? '/' : entry.slug,
+export const getDiscordData = async (userID:string) => {
+  if (discordCache.has(userID)) {
+    return discordCache.get(userID) as {username:string,pfp:string} | undefined;
+  }
```
Comment (cyrus01337) [inline src/pages/[...slug].astro:13]: If fetching the avatar fails i.e when a user doesn't exist, the cache entry will remain `undefined`, which will allow for repeat API calls. `null` could be used to special-case this and avoid the situation entirely.

**Review by cyrus01337 — CHANGES_REQUESTED** [review]
File changes were not auto-formatted, most likely due to lack of editor configuration. This was resolved in the devcontainer instance exclusively, though I've neglected non-devcontainer VS Code instances without realising.

~I plan to resolve that matter separately~ (see #9), in the mean time you should run `<package-manager> run format` and commit the changes.

File: src/pages/[...slug].astro (hunk 3 above)
[the remark this answers is not in the document]
Comment (haotian2006) [reply src/pages/[...slug].astro:8]: for some reason when I run without the export it says it doesn't exists

[the remark this answers is not in the document]
Comment (haotian2006) [reply src/pages/[...slug].astro:8]: and putting an export seems to fix that

[the remark this answers is not in the document]
Comment (haotian2006) [reply src/pages/[...slug].astro:8]: image

[the remark this answers is not in the document]
Comment (cyrus01337) [reply src/pages/[...slug].astro:8]: Strange, I can't seem to reproduce this in the devcontainer setup, try replicating this using the devcontainer setup. Also, what command are you using to run this?

[the remark this answers is not in the document]
Comment (cyrus01337) [reply src/pages/[...slug].astro:8]: When `getStaticPaths` is called, it doesn't take in the top-level/global variables defined. A workaround is hoisting the variables by exporting them from a separate module, which you import in this file, guaranteeing the values are top-level.

I recommend asking in the Astro server what the cause could be as this doesn't make immediate sense why this behaviour would occur.

**Comment by cyrus01337** [comment]
If you're struggling to avoid tracking  `.astro/` by the way, you should be able to run `git rm --cached .astro/` which will tell Git to stop tracking it while keeping the files intact on your filesystem.

**Review by cyrus01337 — APPROVED** [review]
Great work on this. :slightly_smiling_face:

**Pull request merged**